### ISSUE TITLE
🚀 Feature: Catalog extension point to replace default processors or ability to override processors

### ISSUE THREAD

**Issue by manusant** [backstage/backstage#27004]
### 🔖 Feature description

The CatalogBuilder provides 2 functions to support customisations of default catalog processors: `replaceProcessors(processors: CatalogProcessor[])` and `getDefaultProcessors()` that allows us to override the default behaviour of such providers. 

While these capabilities were part of the old backend system and remain accessible in the CatalogBuilder, we propose exposing them through an **extension point**. This would allow developers to easily override or extend processor functionality in a modular way, leveraging the flexibility of the new backend system without directly modifying the core setup.

### 🎤 Context

For GitHub enterprise we found an issue with the default `UrlReaderProcessor` that was removing everything from the catalog if GitHub is down when scanning so, we've implemented a custom UrlReaderProcessor and avoid removing entities in such scenario:

```ts
export class CustomUrlReaderProcessor extends UrlReaderProcessor {
  private readonly logger: Logger;

  constructor(options: Options) {
    super(options);
    this.logger = options.logger;
  }

  async readLocation(
    location: LocationSpec,
    optional: boolean,
    emit: CatalogProcessorEmit,
    parser: CatalogProcessorParser,
    cache: CatalogProcessorCache,
  ): Promise<boolean> {
    const emitter: CatalogProcessorEmit = generated => {
      if (
        generated.type === 'error' &&
        generated.error.message
          ?.toLowerCase()
          .includes('unable to read url, httperror: fetch failed')
      ) {
        this.logger.debug(`Invalid reading request, GHE is down.`);
        return;
      }
      emit(generated);
    };
    return await super.readLocation(location, optional, emitter, parser, cache);
  }
}
```
Then we were replacing the default one by doing:
```ts
const processors = builder
    .getDefaultProcessors()
    .filter((p: unknown) => !(p instanceof UrlReaderProcessor));
  processors.push(
    new CustomUrlReaderProcessor({ reader: env.reader, logger: env.logger }),
  );
  builder.replaceProcessors(processors);
```

We would like to retain this flexibility to customize which processors are available and control their priority. This allows us to handle specific failure conditions (such as external system downtime) without risking the integrity of the catalog, ensuring better reliability and control over the entity processing flow.

### ✌️ Possible Implementation

Doing the new way should be an extension point.

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by Rugvip** [comment]
Yep this is indeed missing. Open to suggestions what the best way to make this available in the new backend system is. Ideally we'd avoid re-exporting the existing built-in processors. Maybe the processing extension point could let one disable specific built-in processors? 🤔

**Comment by malagant** [comment]
I am also currently facing the missing feature from CatalogBuilder which is deprecated. In my case, I just want to remove the `BuiltinKindsEntityProcessor` from the default processors. Wouldn't it be nice to have it like this?

```typescript
import { catalogProcessingExtensionPoint } from '@backstage/plugin-catalog-node/alpha';

export const catalogModuleProcessorConfig = createBackendModule({
  pluginId: 'catalog',
  moduleId: 'processor-config',
  register(env) {
    env.registerInit({
      deps: {
        catalog: catalogProcessingExtensionPoint,
        logger: coreServices.logger,
      },
      async init({ catalog, logger }) {
        catalog.removeProcessor('BuiltinKindsEntityProcessor');
      },
    });
  },
});
```

I'm just loud thinking. ;)

**Comment by Rugvip** [comment]
Poking around at this a bit I realized we already have a bit of logic in place to allow for replacement: https://github.com/backstage/backstage/blob/b4c90d9f1b02017f1186c08c2a899e84c854d82a/plugins/catalog-backend/src/service/CatalogBuilder.ts#L723-L734

Given that that piece of code already exists I think it would be best to rely on that for now, and would be happy to add support for the same kind of overriding for the other built-in processors.

**Comment by chicobento** [comment]
hi all, 

We have a similar case here. We created a processor to support multiple owners via `owners` field.
The processor basically emits the owner relationships and synthetically produces the `owner` field based on the first item of the array.
The problem is that the processor runs after the `CodeOwnersProcessor`, and by the time the `CodeOwnersProcessor` kicks in, it won't find the `owner` property and will do unnecessary round trips to the SCM to find the `CODEOWNERS` file: https://github.com/backstage/backstage/blob/c934c49457e368ebdc81fd8c2517237b59e03eec/plugins/catalog-backend/src/processors/CodeOwnersProcessor.ts#L67-L72.
Removing / reordering the processors would be very appreciated to workaround this.

**Comment by Rugvip** [comment]
Yep that's an interesting point, I think coming back to https://github.com/backstage/backstage/issues/27004#issuecomment-2579805017 it might be best to have that logic replace the processor in-place rather than pushing it to the end of the processors array.

**Comment by chicobento** [comment]
hi @Rugvip , thanks for the reply. IMO, even though the proposal https://github.com/backstage/backstage/issues/27004#issuecomment-2579805017 would work for the short-term, I think that pushing this type of logic to every processor is limited as:

- it relies too much on the processor code provider taste
- it also does not cover the reordering of processors - i.e: when one processor may add a property that the next one may use to trigger its logic. in the case I mentioned above I'd put my processor before `CodeOwnersProcessor` so that it'd still be executed in `owners` are not defined in the entity
 
For the mid-long term I'd personally prefer a generic way of handing over this control to the backstage instance maintainers - as the processors may come from any plugin, not necessarily from the built-in ones - even better if it could be done via configuration, something like:
```yaml
catalog:
  entityPprocessors:
    enable:
      all: true
      CodeOwnersProcessor: false
    order:
      - *
      - MyOwnersProcessor
      - CodeOwnersProcessor
```

**Comment by Rugvip** [comment]
Had a discussion around this in the framework SIG and settled that the best way forward for now is to export the built-in processors and make it possible to replace all of them via an extension point. The exports should be via a sub-path export that does not promise a stable API, so that we can add dependencies to the processors without that being considered a breaking change.

**Comment by punkle** [comment]
@manusant we had a look into this today and found that it is already possible to replace a processor using the addProcessor function. The key is to ensure that the processor you would like to replace is a default process and the new processor has the same name (as returned by getProcessorName) as the processor you want to replace.

e.g. the following would replace the built in url-reader processor.

```typescript
catalog.addProcessor({
  getProcessorName: () => 'url-reader'
})
```

**Comment by punkle** [comment]
I added documentation to explain this point here: https://github.com/backstage/backstage/pull/29469

**Comment by chicobento** [comment]
> [@manusant](https://github.com/manusant) we had a look into this today and found that it is already possible to replace a processor using the addProcessor function. The key is to ensure that the processor you would like to replace is a default process and the new processor has the same name (as returned by getProcessorName) as the processor you want to replace.
> 
> e.g. the following would replace the built in url-reader processor.
> 
> catalog.addProcessor({
>   getProcessorName: () => 'url-reader'
> })

Thank you @punkle , this hack will solve my issue with `CodeOwnersProcessor`

**Comment by Ishubhammohole** [comment]
Hi! I’ve gone through the discussion and would like to work on implementing this feature by exposing an extension point for processor replacement as described above.

From what I understand, the current workaround using `getProcessorName` works, but a proper solution via configuration or the extension system would be more robust long term.

If the team is still open to contributions here, I'd love to take this up and start working on a PR. Please let me know!

**Comment by freben** [comment]
Hi, sorry - after the [disableDefaultProcessors config flag](https://github.com/backstage/backstage/blob/14b25217517ec9ebbe9b3f8204796a6f97e83893/plugins/catalog-backend/config.d.ts#L158) was added in https://github.com/backstage/backstage/pull/29478, this is already achievable. Closing the issue.
